### PR TITLE
make delete faster, use an NIO file walker (patched)

### PR DESCRIPTION
**Pull request by cunei**
See https://github.com/sbt/io/pull/133

### CODE REVIEW

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/515/console">is in progress...</a>

File: io/src/main/scala/sbt/io/IO.scala [hunk 1]
```diff
@@ -454,13 +462,35 @@ object IO {
...
   def delete(file: File): Unit = {
-    translate("Error deleting file " + file + ": ") {
-      val deleted = file.delete()
-      if (!deleted && file.isDirectory) {
-        delete(listFiles(file))
-        file.delete
-        ()
+    object deleter extends SimpleFileVisitor[NioPath] {
+      override def visitFile(file: NioPath, attr: BasicFileAttributes): FileVisitResult = {
+        translate("Error deleting file " + file.toFile + ": ") {
```
Comment (mkeskells) [inline io/src/main/scala/sbt/io/IO.scala]: as translate is a exception wrapper would it be better to have just one wrapper?

Comment (cunei) [reply io/src/main/scala/sbt/io/IO.scala]: It's up to you how to organize the code; I'm just pointing out the reason for the failures described at https://github.com/sbt/io/pull/133#pullrequestreview-99974172.

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-1.x-pull/135/head-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/515/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@90cd60f3b90a1b640453d8f3988a904f49ce2d19
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/1.x">1.x                 </A> | sbt/zinc@ad46221ca148629df111dbc1561ac348d2b15c9c
  io                | <A HREF="https://github.com/sbt/io/pull/135">pull/135/head       </A> | sbt/io@6b2d5b73ea33b650494fc156b9b80d3bc9caf051
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@0d21ae6369a3e4054ed434902ea84b2520d21da9
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@85a30fac4a3f895c5aa6571936bc7d016a0c962c
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=1.x&io=pull/135/head&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=1.x&zinc=1.x&io=pull/135/head&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

File: io/src/main/scala/sbt/io/IO.scala [hunk 2]
```diff
@@ -454,13 +462,35 @@ object IO {
 
   /** Deletes `file`, recursively if it is a directory. */
   def delete(file: File): Unit = {
-    translate("Error deleting file " + file + ": ") {
```
Comment (eed3si9n) [inline io/src/main/scala/sbt/io/IO.scala:457]: The entire delete should be wrapped in `translate`, as well as the individual walking.

**Review by eed3si9n — CHANGES_REQUESTED** [review]

**Comment by cunei** [comment]
The `translate` wrapper is now improved. In case of an error deleting a nested file in a directory tree (other than file not found), the final exception text will list the root of the directory tree, as well as the file delete operation that failed, as in:
```
[error] Error during a recursive delete of path /tmp/a: Error deleting /tmp/a/b/sdkfas: java.nio.file.AccessDeniedException: /tmp/a/b/sdkfas
```

**Comment by cunei** [comment]
I modified https://github.com/sbt/io/pull/135 according to Eugene's instructions.
A couple of considerations:
1) the original `IO.delete()` will swallow not only the `NoSuchFileException`, but all the IO errors. If a file cannot be delete because it belongs to a different user, for example, the old `IO.delete` will blissfully say nothing, and the file will not actually be deleted. I think continuing to ignore the `NoSuchFileException`s is fine, as long as we take the meaning of `IO.delete()` to be "file-be-gone." However, the new code will actually catch IO exceptions that the old code would ignore. I would call this an improvement, actually.
2) The underlying cause of the troubles with `IO.delete` is that multiple recursive delete operations may take place at the same time. Ignoring `NoSuchFileException`s is a cheap way to sidestep the potential conflicts, but at least potentially it may happen that concurrent file traversals and deletions interfere in undesirable ways. A more permanent fix would be to rename (atomically if possible) the directory that should be deleted to something else, and then deleting without fear of conflicts. `Files` has a handy:
`Files.move(path, targetPath, ATOMIC_MOVE)` which we could use in the future, if we ever wanted to make this code more robust.

Besides these considerations, I think this pull request can be merged at this point. I restarted the validator, to double check.

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/566/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-1.x-pull/135/head-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/566/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@ca874d5d5f6046c4ac2d4825d81e84b97b0f1c0a
  zinc              | <A HREF="https://github.com/sbt/zinc/tree/1.x">1.x                 </A> | sbt/zinc@d6dec01ca95a21fc11f62669b44f1ed4b28ae42d
  io                | <A HREF="https://github.com/sbt/io/pull/135">pull/135/head       </A> | sbt/io@3b00df3478d9b3cb756a6abb476e0055986756b0
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@c6b2b626c8e37de40a8df99ef89d17dbf54bf311
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@85a30fac4a3f895c5aa6571936bc7d016a0c962c
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=1.x&io=pull/135/head&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Review by dwijnand — APPROVED** [review]
thank you!

**Pull request merged**